File: DataStructures/Trees/BinaryIndexedTree/FenwickTree/atcoder/A76/A76.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

const MOD = 1_000_000_007

type FenwickTree struct {
	n    int
	data []int
}

func NewFenwickTree(n int) *FenwickTree {
	return &FenwickTree{
		n:    n,
		data: make([]int, n+2),
	}
}

func (ft *FenwickTree) Add(i, x int) {
	i++
	for i <= ft.n+1 {
		ft.data[i] = (ft.data[i] + x) % MOD
		i += i & -i
	}
}

func (ft *FenwickTree) Sum(i int) int {
	i++
	res := 0
	for i > 0 {
		res = (res + ft.data[i]) % MOD
		i -= i & -i
	}
	return res
}

func (ft *FenwickTree) RangeSum(l, r int) int {
	return (ft.Sum(r) - ft.Sum(l-1) + MOD) % MOD
}

func lowerBound(a []int, x int) int {
	return sort.Search(len(a), func(i int) bool { return a[i] >= x })
}

func upperBound(a []int, x int) int {
	return sort.Search(len(a), func(i int) bool { return a[i] > x })
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords) // 単語（スペース区切り）単位で読み取り

	readInt := func() int {
		scanner.Scan()
		n, _ := strconv.Atoi(scanner.Text())
		return n
	}

	N := readInt()
	W := readInt()
	L := readInt()
	R := readInt()

	X := make([]int, N)
	for i := 0; i < N; i++ {
		X[i] = readInt()
	}

	// 全地点（スタート0、足場、ゴールW）をリストアップ
	positions := append([]int{0}, X...)
	positions = append(positions, W)
	sort.Ints(positions)

	posIndex := make(map[int]int)
	for i, v := range positions {
		posIndex[v] = i
	}

	n := len(positions)
	dp := make([]int, n)
	dp[0] = 1

	ft := NewFenwickTree(n)
	ft.Add(0, 1)

	for i := 1; i < n; i++ {
		cur := positions[i]
		left := cur - R
		right := cur - L

		li := lowerBound(positions, left)
		ri := upperBound(positions, right) - 1

		if li <= ri {
			val := ft.RangeSum(li, ri)
			dp[i] = val
			ft.Add(i, val)
		}
	}

	fmt.Println(dp[n-1])
}

// Go 解法で 20件の入力で ランタイムエラー（panic） が発生するとのこと、原因として考えられるのは以下のいずれかです：

// ❗️ 原因候補
// bufio.Scanner の 2行しか読んでいない
// 実際には N が大きくなると 複数行に分かれて入力される 可能性がある。
// つまり "5 65 7 37" の次に "5 15 30 50 55" でなく、改行を含む複数行に X が分割されている可能性がある。
// Xstr := strings.Fields(scanner.Text()) だけで全 N 要素を取得できない。

// ✅ 修正方針
// N 個の X[i] を読み切るまで ループで scanner.Scan() を繰り返す
// X := make([]int, N) を安全に埋める

// ✅ 修正点まとめ
// 修正箇所	内容
// scanner.Split(...)	単語単位で int を逐次読み取り可能に
// readInt() 関数	scanner.Scan() + strconv.Atoi() のラッパー
// X[i] = readInt()	必ず N 件読み切るようループで読み取り

File: DataStructures/Trees/BinaryIndexedTree/FenwickTree/atcoder/A76/A76.ts
import * as fs from 'fs';

const MOD = 1_000_000_007;

class FenwickTree {
    private n: number;
    private tree: number[];

    constructor(n: number) {
        this.n = n;
        this.tree = new Array(n + 2).fill(0);
    }

    add(i: number, x: number): void {
        i++;
        while (i <= this.n + 1) {
            this.tree[i] = (this.tree[i] + x) % MOD;
            i += i & -i;
        }
    }

    sum(i: number): number {
        i++;
        let res = 0;
        while (i > 0) {
            res = (res + this.tree[i]) % MOD;
            i -= i & -i;
        }
        return res;
    }

    rangeSum(l: number, r: number): number {
        return (this.sum(r) - this.sum(l - 1) + MOD) % MOD;
    }
}

function lowerBound(arr: number[], x: number): number {
    let left = 0,
        right = arr.length;
    while (left < right) {
        const mid = (left + right) >> 1;
        if (arr[mid] < x) left = mid + 1;
        else right = mid;
    }
    return left;
}

function upperBound(arr: number[], x: number): number {
    let left = 0,
        right = arr.length;
    while (left < right) {
        const mid = (left + right) >> 1;
        if (arr[mid] <= x) left = mid + 1;
        else right = mid;
    }
    return left;
}

// Main function
function main(): void {
    const input = fs.readFileSync('/dev/stdin', 'utf8');
    const lines = input.trim().split('\n');

    const [N, W, L, R] = lines[0].split(' ').map(Number);
    const X = lines[1].split(' ').map(Number);

    // 全地点（0, 足場, W）を座標圧縮用にソート
    const positions = [0, ...X, W];
    positions.sort((a, b) => a - b);

    const posToIndex = new Map<number, number>();
    positions.forEach((v, i) => posToIndex.set(v, i));

    const n = positions.length;
    const dp = new Array(n).fill(0);
    dp[0] = 1;

    const ft = new FenwickTree(n);
    ft.add(0, 1);

    for (let i = 1; i < n; i++) {
        const cur = positions[i];
        const left = cur - R;
        const right = cur - L;

        const li = lowerBound(positions, left);
        const ri = upperBound(positions, right) - 1;

        if (li <= ri) {
            const val = ft.rangeSum(li, ri);
            dp[i] = val;
            ft.add(i, val);
        }
    }

    console.log(dp[n - 1]);
}

main();

// 📌 解法ポイント
// 座標圧縮：Wが大きすぎる（最大10⁹）ので使う点だけに圧縮
// DP + BIT：dp[i] = Σ dp[j], 条件：L ≤ pos[i] - pos[j] ≤ R → これを BIT で高速集計
// 計算量：O(N log N) で N ≤ 150,000 に対応

